modules/product/src/Tests/AddToCartFormTest.php

```diff
@@ -0,0 +1,342 @@
+    $product = $this->variation->getProduct();
+
+    /**
+     * +--------------------+------------------+
+     * | Variation          | Attributes       |
+     * +--------------------+------------------+
+     * | 1                  | Small, Red       |
+     * | 2                  | Medium, Red      |
+     * | 3                  | Medium, Blue     |
+     * +--------------------+------------------+
+     */
+
+    // Update first variation to have the attribute's value.
+    $this->variation->set('test_size_attribute', $size_attributes['small']->id());
+    $this->variation->set('test_color_attribute', $color_attributes['red']->id());
+    $this->variation->save();
+
+    $variation2 = $this->createEntity('commerce_product_variation', [
+      'type' => $variation_type->id(),
+      'sku' => $this->randomMachineName(),
+      'price' => [
+        'amount' => 999,
+        'currency_code' => 'USD',
+      ],
+      'test_size_attribute' => $size_attributes['medium']->id(),
+      'test_color_attribute' => $color_attributes['red']->id(),
+    ]);
+    $variation2->save();
+
+    $variation3 = $this->createEntity('commerce_product_variation', [
+      'type' => $variation_type->id(),
+      'sku' => $this->randomMachineName(),
+      'price' => [
+        'amount' => 999,
+        'currency_code' => 'USD',
+      ],
+      'test_size_attribute' => $size_attributes['medium']->id(),
+      'test_color_attribute' => $color_attributes['blue']->id(),
+    ]);
+    $variation3->save();
+
+    $product->variations->appendItem($variation2);
+    $product->variations->appendItem($variation3);
+    $product->save();
+
+
+    $this->postAddToCart($this->variation->getProduct());
+    // Trigger AJAX by changing size attribute
+    $ajax_commands = $this->drupalPostAjaxForm(NULL, [
+      'attributes[test_size_attribute]' => $size_attributes['medium']->id(),
+    ], 'attributes[test_size_attribute]');
+    // Trigger AJAX by changing color attribute
+    $ajax_commands = $this->drupalPostAjaxForm(NULL, [
+      'attributes[test_color_attribute]' => $color_attributes['blue']->id(),
+    ], 'attributes[test_color_attribute]');
+
+    // We can't assert an option doesn't exist using AssertContentTrait, since
+    // our ID is dynamic. Version of assertNoOption using data-drupal-selector.
+    // @see \Drupal\simpletest\AssertContentTrait::assertNoOption
+    $selects = $this->xpath('//select[@data-drupal-selector=:data_drupal_selector]', [
+      ':data_drupal_selector' => 'edit-attributes-test-size-attribute',
+    ]);
+    $options = $this->xpath('//select[@data-drupal-selector=:data_drupal_selector]//option[@value=:option]', [
+      ':data_drupal_selector' => 'edit-attributes-test-size-attribute',
+      ':option' => $size_attributes['small']->id(),
+    ]);
+    $this->assertTrue(isset($selects[0]) && !isset($options[0]), NULL, 'Browser');
+
+    // Since we do not have a Small, Blue. Should only see variation 3.
```

Comment:
Since Blue is the only option, the form element should be #disabled, can we confirm that?